eloquentjs/7_2_robot_efficiency.js

```diff
@@ -0,0 +1,463 @@
+}
+
+
+function randomPick(array) {
+    const choice = Math.floor(Math.random() * array.length);
+    return array[choice];
+}
+
+function findRoute(graph, from, to) {
+    const work = [{at: from, route: []}];
+
+    for (let i = 0; i < work.length; i++) {
+      const {at, route} = work[i];
+
+      for (let place of graph[at]) {
+        if (place === to) {
+            return route.concat(place);
+        }
+
+        if (!work.some(w => w.at === place)) {
+          work.push({at: place, route: route.concat(place)});
+        }
+      }
+    }
+}
+
+/**
+ * ----------- ROBOT SIMULATION/TESTING -----------
+ */
+
+function runRobot(state, robot, memory) {
+    const maxIter = 5000;
+    for (let turn = 0; turn < maxIter; turn++) {
+        if (state.parcels.length === 0) {
+            return turn;
+        }
+
+        let action = robot(state, memory);
+        state =  state.move(action.direction);
+        memory = action.memory;
+    }
+    throw `${robot.name} HAS EXCEEDED ${maxIter} ITERATIONS`;
+}
+
+function robotComparison(robots, numTests, parcelCount) {
+    const tests = [];
+    for (let i = 0; i < numTests; i++) {
+        tests.push(new VillageState.random(parcelCount))
+    }
+
+    const results = [];
+    for (robot of Object.keys(robots)) {
+        console.log(`Testing: ${robot}`);
+
+        let acc = 0;
+        const t0 = performance.now();
+        for (test of tests) {
+            acc += runRobot(test, robots[robot]);
+        }
+        const t1 = performance.now();
+        const delta = t1 - t0;
+        
+        results.push({name: robot, averageSteps: acc / numTests, averageMilliseconds: delta / numTests});
+    }
+
+    return results;
+}
+
+/**
+ * ----------- ROBOT ZOO (book)-----------
+ */
+
+function randoBot(state) {
+    return {direction: randomPick(roadGraph[state.place])}
+}
+
+function routeBot(state, memory = []) {
+    const mailRoute = [
+        "Alice's House", "Cabin", "Alice's House", "Bob's House",
+        "Town Hall", "Daria's House", "Ernie's House",
+        "Grete's House", "Shop", "Grete's House", "Farm",
+        "Marketplace", "Post Office"
+    ];
+    
+    if (memory.length === 0) {
+      memory = mailRoute;
+    }
+    
+    return {direction: memory[0], memory: memory.slice(1)};
+}
+
+function goalOrientedBot({place, parcels}, route = []) {
+    if (route.length === 0) {
+      const parcel = parcels[0];
+
+      if (parcel.place !== place) {
+        route = findRoute(roadGraph, place, parcel.place);
+      } else {
+        route = findRoute(roadGraph, place, parcel.address);
+      }
+    }
+    
+    return {direction: route[0], memory: route.slice(1)};
+}
+
+/**
+ * ----------- ROBOT ZOO (originals) -----------
+ */
+
+/**
+ * First uses brute force search the find the shortest route and then follows it
+ */
+function brutusBot(state,  memory = {route: []}) {
+    let bestRoute;
+    let relevantLocations = [];
+    let shortestRouteLengthEst = Infinity;
```

Comment:
My thinking with the estimate is that by first running runRobot with a more performative bot that finds a 'pretty good' path, it places an upper bound on the shortest path. The upper bound allows brutus to later return early on from paths that are by definition not the shortest path because its already been beaten by the cheaper bot. In hindsight I think this should be named `shortestRouteUpperBound`.